### PR TITLE
feat(core): SiteContentGraph actor (#136)

### PR DESCRIPTION
**Pull request by davidwkeith**
## Summary

- Lands issue #136 — `SiteContentGraph` actor in `AnglesiteCore`, the foundation projection for Siri AI Phase A.
- Push-driven, single-subscriber change handler keyed by siteID; subscriber (A.3 `ContentSpotlightIndexer`, issue #144) reads back for its diff — same pattern as the existing `SpotlightIndexer.reindex(_:)`.
- No-op suppression on `upsert*` via `Equatable` so file-watch fires with identical content do not wake the indexer.
- 23 Swift Testing cases covering bulk-load replacement, siteID isolation (across pages/posts/images), upsert/remove emit semantics, no-op suppression, unload (always-emit even when empty), case-insensitive search across the spec'd fields (title/route for pages; title/slug/tags/collection for posts), empty-query convention, `knownSiteIDs` enumeration, handler detach, and actor-isolation serialization under 100-way TaskGroup parallelism.

Design: [`docs/superpowers/specs/2026-06-11-site-content-graph-design.md`](docs/superpowers/specs/2026-06-11-site-content-graph-design.md) · Plan: [`docs/superpowers/plans/2026-06-11-site-content-graph.md`](docs/superpowers/plans/2026-06-11-site-content-graph.md)

## Test plan

- [x] `swift test --package-path .` — 23/23 `SiteContentGraphTests` pass. Two pre-existing baseline failures (`MCPClientHTTPEndToEndTests`, `AppliesEditEndToEndTests`) are unrelated to this branch — both are MCP-transport e2e tests gated on plugin/node infra, fail identically on `main`.
- [x] `xcodebuild -scheme Anglesite -configuration Debug build` — **BUILD SUCCEEDED**
- [x] `xcodebuild -scheme AnglesiteMAS -configuration Debug build` — **BUILD SUCCEEDED**

## What this PR does NOT do

Per the spec's scope boundaries:
- No MCP / plugin wiring (A.8, #142).
- No `IndexedEntity` conformance on Page/Post/Image (A.2, #137).
- No `ContentSpotlightIndexer` (A.3, #144).
- No `list_content` MCP tool (plugin paired PR, #140 / A.6).
- No persistence to disk (in-memory by design).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: Sources/AnglesiteCore/SiteContentGraph.swift [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        for key in self.posts.compactMap({ $0.value.siteID == siteID ? $0.key : nil }) {
+            self.posts.removeValue(forKey: key)
+        }
+        for key in self.images.compactMap({ $0.value.siteID == siteID ? $0.key : nil }) {
+            self.images.removeValue(forKey: key)
+        }
+        for page in pages { self.pages[page.id] = page }
+        for post in posts { self.posts[post.id] = post }
+        for image in images { self.images[image.id] = image }
+        await emitChange(siteID)
```
Comment (davidwkeith) [inline Sources/AnglesiteCore/SiteContentGraph.swift:147]: **Bug: `load()` silently accepts items whose `.siteID` doesn't match the `siteID` parameter, poisoning cross-site isolation.**

Lines 144–146 insert every item keyed by `item.id` without checking `item.siteID == siteID`. The eviction loop on lines 135–143 only removes entries where `value.siteID == siteID`, so a rogue item with `.siteID == "B"` passed to `load(siteID: "A", …)` survives and pollutes `pages(for: "B")` until a subsequent `load`/`unload` for site B cleans it up. This violates Invariant 1 (siteID isolation).

Add a guard before inserting:
```swift
assert(pages.allSatisfy { $0.siteID == siteID }, "load: page siteID mismatch")
assert(posts.allSatisfy { $0.siteID == siteID }, "load: post siteID mismatch")
assert(images.allSatisfy { $0.siteID == siteID }, "load: image siteID mismatch")
```
Or silently filter and log a fault via `LogCenter`. Either way a test is needed for this path (see test-file comment below).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Sources/AnglesiteCore/SiteContentGraph.swift [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    ) async {
+        for key in self.pages.compactMap({ $0.value.siteID == siteID ? $0.key : nil }) {
+            self.pages.removeValue(forKey: key)
+        }
+        for key in self.posts.compactMap({ $0.value.siteID == siteID ? $0.key : nil }) {
+            self.posts.removeValue(forKey: key)
+        }
+        for key in self.images.compactMap({ $0.value.siteID == siteID ? $0.key : nil }) {
+            self.images.removeValue(forKey: key)
+        }
```
Comment (davidwkeith) [inline Sources/AnglesiteCore/SiteContentGraph.swift:143]: **Nit/perf: two-pass eviction (`compactMap` + N × `removeValue`) when a single filter suffices.**

The shipped code allocates an intermediate `[String]` of keys then removes them one by one — O(n) extra allocations. The cleaner and faster form is:

```swift
self.pages  = self.pages.filter  { $0.value.siteID != siteID }
self.posts  = self.posts.filter  { $0.value.siteID != siteID }
self.images = self.images.filter { $0.value.siteID != siteID }
```

Same issue exists in `unload()` at lines 210–218. The plan doc (Task 3 Step 3, Task 7 Step 3) shows this single-pass form.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Sources/AnglesiteCore/SiteContentGraph.swift [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    public func unload(siteID: String) async {
+        for id in pages.compactMap({ $0.value.siteID == siteID ? $0.key : nil }) {
+            pages.removeValue(forKey: id)
+        }
+        for id in posts.compactMap({ $0.value.siteID == siteID ? $0.key : nil }) {
+            posts.removeValue(forKey: id)
+        }
+        for id in images.compactMap({ $0.value.siteID == siteID ? $0.key : nil }) {
+            images.removeValue(forKey: id)
+        }
```
Comment (davidwkeith) [inline Sources/AnglesiteCore/SiteContentGraph.swift:218]: Same two-pass eviction issue as in `load()` above. Prefer:

```swift
pages  = pages.filter  { $0.value.siteID != siteID }
posts  = posts.filter  { $0.value.siteID != siteID }
images = images.filter { $0.value.siteID != siteID }
```

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Sources/AnglesiteCore/SiteContentGraph.swift [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+        let scoped = posts.values.filter { $0.siteID == siteID }
+        guard !query.isEmpty else { return Array(scoped) }
+        let needle = query.lowercased()
+        return scoped.filter { post in
+            if post.title.lowercased().contains(needle) { return true }
+            if post.slug.lowercased().contains(needle) { return true }
+            if post.collection.lowercased().contains(needle) { return true }
+            if post.tags.contains(where: { $0.lowercased().contains(needle) }) { return true }
+            return false
+        }
```
Comment (davidwkeith) [inline Sources/AnglesiteCore/SiteContentGraph.swift:255]: **`lowercased()` is not locale-aware — broken for Turkish, Azerbaijani, and other locales with special case folding.**

`"İstanbul".lowercased()` produces different results under `.current` vs a hardcoded locale. For user-supplied titles, slugs, and routes, use `range(of:options:locale:)` with `.caseInsensitive` instead:

```swift
func matches(_ s: String) -> Bool {
    s.range(of: query, options: .caseInsensitive, locale: .current) != nil
}
```

This removes the need to pre-lowercase `needle` at all, and applies equally to `searchPages` and `searchPosts`. Same issue at lines 248–253 in `searchPosts`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Sources/AnglesiteCore/SiteContentGraph.swift [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+    private var pages: [String: Page] = [:]
+    private var posts: [String: Post] = [:]
+    private var images: [String: Image] = [:]
+    private var changeHandler: ChangeHandler?
+
+    public init() {}
+
+    public func setChangeHandler(_ handler: ChangeHandler?) {
+        changeHandler = handler
+    }
```
Comment (davidwkeith) [inline Sources/AnglesiteCore/SiteContentGraph.swift:118]: **Missing doc comment on reentrancy semantics.**

A caller who installs a handler, fires an upsert, then immediately calls `setChangeHandler(nil)` from a concurrent task may reasonably ask: does the old handler still fire? Under Swift's actor model the answer is yes — the actor serialises both calls so the nil write takes effect after any suspended `emitChange` resumes. This is correct but non-obvious. Add a brief comment:

```swift
/// Replaces (or removes, when `nil`) the current handler. Safe to call from any isolation
/// context. Does not cancel an in-flight handler call — a concurrent `setChangeHandler(nil)`
/// takes effect after the current `emitChange` suspension point returns.
```

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Tests/AnglesiteCoreTests/SiteContentGraphTests.swift [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+
+    @Test("Page is Equatable, Identifiable, and id contains siteID + route")
+    func pageStructShape() {
+        let p1 = Self.page()
+        let p2 = Self.page()
+        let p3 = Self.page(route: "/contact")
+        #expect(p1 == p2)
+        #expect(p1 != p3)
+        #expect(p1.id == "\(Self.siteA):page:/about")
+        #expect(p1.id == p1.id) // exists as Identifiable.ID
```
Comment (davidwkeith) [inline Tests/AnglesiteCoreTests/SiteContentGraphTests.swift:90]: **Tautological assertion — always passes regardless of implementation.**

`#expect(p1.id == p1.id)` is `x == x`; it can never fail. The intent seems to be "verify the `id` conforms to `Identifiable.ID`", but that's a compile-time structural guarantee, not a runtime one. The comment "`// exists as Identifiable.ID`" adds noise rather than signal.

Remove the line, or replace with something that can actually catch a regression — e.g. `#expect(p1.id.contains(Self.siteA))` to verify the siteID is embedded in the `id`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Tests/AnglesiteCoreTests/SiteContentGraphTests.swift [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+        modified: Date = referenceDate
+    ) -> SiteContentGraph.Post {
+        SiteContentGraph.Post(
+            id: "\(site):post:\(slug)",
+            siteID: site,
+            collection: collection,
+            slug: slug,
+            title: title,
+            draft: draft,
+            publishDate: modified,
```
Comment (davidwkeith) [inline Tests/AnglesiteCoreTests/SiteContentGraphTests.swift:44]: **`publishDate` always equals `lastModified` in the fixture — the two fields are indistinguishable in every test.**

`Post` has distinct semantics for `publishDate` (editorial date) and `lastModified` (filesystem mtime). Setting them to the same value means no test exercises a `publishDate`-only change, and no test confirms that such a change triggers an emit (it would, because `Equatable` includes both fields). Consider:

```swift
publishDate: modified.addingTimeInterval(-86400),  // one day before lastModified
lastModified: modified
```

Then add a test that upserts a post where only `publishDate` changed and verifies the change is emitted.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Tests/AnglesiteCoreTests/SiteContentGraphTests.swift [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+        let bPages = await graph.pages(for: Self.siteB)
+        let bPosts = await graph.posts(for: Self.siteB)
+        let bImages = await graph.images(for: Self.siteB)
+
+        #expect(aPages.map(\.route) == ["/a-home"])
+        #expect(aPosts.map(\.slug) == ["a-post"])
+        #expect(aImages.map(\.relativePath) == ["public/images/a.jpg"])
+        #expect(bPages.map(\.route) == ["/b-home"])
+        #expect(bPosts.map(\.slug) == ["b-post"])
+        #expect(bImages.map(\.relativePath) == ["public/images/b.jpg"])
```
Comment (davidwkeith) [inline Tests/AnglesiteCoreTests/SiteContentGraphTests.swift:141]: **Ordered array equality on results sourced from an unordered dictionary — latent flakiness.**

`pages.values`, `posts.values`, and `images.values` have no guaranteed iteration order; it varies across Swift versions, OS builds, and hash randomisation. These assertions pass today only because each result set happens to have a single element (single-element ordered == unordered). Any future test that adds a second matching item and expects a specific order will be flaky.

Adopt the same pattern `searchWithEmptyQueryReturnsAll` uses (lines 425–426): `Set(results.map(\.route))`, or call `.sorted()` on a stable property before comparing, like `loadReplacesExistingEntries` does at line 110. Apply consistently to lines 136–141, 348, 383–384, and 407–410.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Tests/AnglesiteCoreTests/SiteContentGraphTests.swift [hunk 9]
```diff
@@ -0,0 +1,476 @@
...
+    func removePageUnknownIdSilent() async {
+        let graph = SiteContentGraph()
+        let counter = TestCounter()
+        await graph.setChangeHandler { siteID in await counter.record(siteID) }
+
+        await graph.removePage(id: "nonexistent:page:/whatever")
+
+        let count = await counter.count
+        #expect(count == 0)
+    }
```
Comment (davidwkeith) [inline Tests/AnglesiteCoreTests/SiteContentGraphTests.swift:289]: **`removePost` and `removeImage` have no "silent no-op for unknown id" test.**

`removePage` has `removePageUnknownIdSilent` here verifying Invariant 4. The identical code paths in `removePost` and `removeImage` are unguarded — a regression that accidentally adds an emit for the unknown-id case in either would not be caught. Add:

```swift
@Test("removePost is silent and no-op when id is unknown")
func removePostUnknownIdSilent() async { … }

@Test("removeImage is silent and no-op when id is unknown")
func removeImageUnknownIdSilent() async { … }
```

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Tests/AnglesiteCoreTests/SiteContentGraphTests.swift [hunk 10]
```diff
@@ -0,0 +1,476 @@
...
+        await graph.setChangeHandler { siteID in await counter.record(siteID) }
+
+        let revised = Self.page(title: "About Us")  // same route -> same id
+        await graph.upsertPage(revised)
+
+        let stored = await graph.page(id: original.id)
+        let count = await counter.count
+        #expect(stored?.title == "About Us")
+        #expect(count == 1)
+    }
```
Comment (davidwkeith) [inline Tests/AnglesiteCoreTests/SiteContentGraphTests.swift:207]: **`upsertPost` and `upsertImage` have no "same id, different value, emits and overwrites" test.**

`upsertPageDifferentValueEmits` (this test) covers the critical mutation path for pages. The equivalent tests are absent for posts and images — a regression in their mutation logic (e.g. an early return that accidentally suppresses the overwrite) would not be caught. Add:

```swift
@Test("upsertPost with same id but different title emits and overwrites")
func upsertPostDifferentValueEmits() async { … }

@Test("upsertImage with same id but different byteSize emits and overwrites")
func upsertImageDifferentValueEmits() async { … }
```

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Tests/AnglesiteCoreTests/SiteContentGraphTests.swift [hunk 11]
```diff
@@ -0,0 +1,476 @@
...
+        let counter = TestCounter()
+        await graph.setChangeHandler { siteID in await counter.record(siteID) }
+
+        let page = Self.page()
+        await graph.upsertPage(page)
+
+        let stored = await graph.page(id: page.id)
+        let count = await counter.count
+        #expect(stored == page)
+        #expect(count == 1)
```
Comment (davidwkeith) [inline Tests/AnglesiteCoreTests/SiteContentGraphTests.swift:174]: **Emitted `siteID` is only asserted in 2 of the 13+ tests that use `TestCounter`.**

`loadEmitsChange` and `unloadAlwaysEmitsChange` both check `counter.lastSiteID`. Every other test that wires up a `TestCounter` only checks `count`. A bug that emits with the wrong siteID (wrong variable, copy-paste error in a future edit) would go undetected for `upsertPage`, `upsertPost`, `upsertImage`, `removePage`, `removePost`, and `removeImage`. Add `#expect(await counter.lastSiteID == Self.siteA)` to all "emits change" variants.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Tests/AnglesiteCoreTests/SiteContentGraphTests.swift [hunk 12]
```diff
@@ -0,0 +1,476 @@
...
+        let afterFirst = await counter.count
+
+        await graph.load(siteID: Self.siteA, pages: [Self.page()], posts: [], images: [])
+        let afterSecond = await counter.count
+        let last = await counter.lastSiteID
+
+        #expect(afterFirst == 1)
+        #expect(afterSecond == 2)
+        #expect(last == Self.siteA)
+    }
```
Comment (davidwkeith) [inline Tests/AnglesiteCoreTests/SiteContentGraphTests.swift:160]: **Missing test: `load()` with an empty payload on a never-loaded siteID must still emit.**

The "always emits" guarantee for `load()` is tested here with a non-empty payload. But the most likely regression point is an `emitChange` call accidentally moved inside a conditional. Calling `load(siteID: X, pages: [], posts: [], images: [])` on a fresh graph should still emit once. Add:

```swift
@Test("load with empty payload on a never-loaded siteID still emits")
func loadEmptyPayloadEmits() async {
    let graph = SiteContentGraph()
    let counter = TestCounter()
    await graph.setChangeHandler { siteID in await counter.record(siteID) }
    await graph.load(siteID: Self.siteA, pages: [], posts: [], images: [])
    #expect(await counter.count == 1)
    #expect(await counter.lastSiteID == Self.siteA)
}
```

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Tests/AnglesiteCoreTests/SiteContentGraphTests.swift [hunk 13]
```diff
@@ -0,0 +1,476 @@
...
+        await graph.load(
+            siteID: Self.siteA,
+            pages: [Self.page(route: "/about")],
+            posts: [],
+            images: []
+        )
+
+        let pages = await graph.pages(for: Self.siteA)
+        #expect(pages.map(\.route).sorted() == ["/about"])
+    }
```
Comment (davidwkeith) [inline Tests/AnglesiteCoreTests/SiteContentGraphTests.swift:111]: **Missing test: `load()` should reject (or at least not silently store) items whose `.siteID` doesn't match the `siteID` parameter.**

See the bug comment on `SiteContentGraph.swift` lines 135–147. Until the production code is fixed, a test here documents the intended contract and serves as a regression guard:

```swift
@Test("load ignores pages whose siteID does not match the siteID parameter")
func loadRejectsItemsWithWrongSiteID() async {
    let graph = SiteContentGraph()
    let wrongSite = Self.page(site: Self.siteB, route: "/b-page")
    await graph.load(siteID: Self.siteA, pages: [wrongSite], posts: [], images: [])
    let bPages = await graph.pages(for: Self.siteB)
    #expect(bPages.isEmpty, "page with siteID B must not appear under site B after loading into site A")
}
```

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by davidwkeith — COMMENTED** [review]
11 issues across the production actor and its test suite. Grouped by priority:

**Must fix (correctness bugs):**
- `load()` silently accepts items whose `.siteID` doesn't match the `siteID` parameter, poisoning cross-site isolation (Invariant 1). See inline comment on lines 135–147.
- `searchPages`/`searchPosts` use `lowercased()` which is not locale-aware — broken for Turkish, Azerbaijani, and others. Use `range(of:options:locale:)` with `.caseInsensitive` instead.

**Should fix (test correctness):**
- Ordered array equality on results from `Dictionary.values` — iteration order is undefined, currently passing only because every affected result set has exactly one element. See lines 136–141, 348, 383–384, 407–410.
- Emitted `siteID` is only asserted in 2 of the 13+ tests using `TestCounter` — a wrong-siteID regression in any upsert/remove would go undetected.
- Tautological assertion `#expect(p1.id == p1.id)` at line 90.

**Test gaps (missing coverage):**
- `removePost` and `removeImage` have no "silent no-op for unknown id" test (unlike `removePage`).
- `upsertPost` and `upsertImage` have no "same id, different value, emits and overwrites" test (unlike `upsertPage`).
- `load()` with an empty payload on a never-loaded siteID has no emit-always test.
- No test for the siteID-mismatch data-poisoning scenario in `load()`.
- `publishDate` always equals `lastModified` in the `post()` fixture — `publishDate`-only mutations are never exercised.

**Nit:**
- `load()` and `unload()` use a verbose two-pass eviction (`compactMap` + N × `removeValue`) where a single `filter` assignment suffices — see inline comments.
- `setChangeHandler` needs a brief doc comment explaining the reentrancy semantics under concurrent nil-replacement.

---
_Generated by [Claude Code](https://claude.ai/code)_
